nexus/db-queries/src/db/datastore/snapshot.rs

```diff
@@ -43,28 +48,127 @@ impl DataStore {
         let gen = snapshot.gen;
         opctx.authorize(authz::Action::CreateChild, authz_project).await?;
 
-        use db::schema::snapshot::dsl;
+        #[derive(Debug, thiserror::Error)]
+        pub enum CustomError {
+            #[error("Resource already exists")]
+            ResourceAlreadyExists,
+
+            #[error("saw AsyncInsertError")]
+            InsertError(AsyncInsertError),
+        }
+
+        type TxnError = TransactionError<CustomError>;
+
+        let snapshot_name = snapshot.name().to_string();
         let project_id = snapshot.project_id;
-        let snapshot: Snapshot = Project::insert_resource(
-            project_id,
-            diesel::insert_into(dsl::snapshot)
-                .values(snapshot)
-                .on_conflict((dsl::project_id, dsl::name))
-                .filter_target(dsl::time_deleted.is_null())
-                .do_update()
-                .set(dsl::time_modified.eq(dsl::time_modified)),
-        )
-        .insert_and_get_result_async(self.pool_authorized(opctx).await?)
-        .await
-        .map_err(|e| match e {
-            AsyncInsertError::CollectionNotFound => Error::ObjectNotFound {
-                type_name: ResourceType::Project,
-                lookup_type: LookupType::ById(project_id),
-            },
-            AsyncInsertError::DatabaseError(e) => {
-                public_error_from_diesel_pool(e, ErrorHandler::Server)
-            }
-        })?;
+
+        let snapshot: Snapshot = self
```

Comment:
(Looking at the old `Project::insert_resource` code) Could we do this without the transaction? `diesel::insert_into` should return the inserted row, which goes through `insert_and_get_result_async`.

It seems possible to call either:

- [insert_and_get_result_async](https://github.com/oxidecomputer/omicron/blob/9818f6fcf5f3f0019c1571e1bec9d3375d8867b2/nexus/db-queries/src/db/collection_insert.rs#L191), and check for "not found" on the result, or to call
- [insert_and_get_results_async](https://github.com/oxidecomputer/omicron/blob/9818f6fcf5f3f0019c1571e1bec9d3375d8867b2/nexus/db-queries/src/db/collection_insert.rs#L213C18-L213C46), and to check for an "empty vec" as the result (aka, nothing was inserted).

Something like:

```rust
       let snapshots: Vec<Snapshot> = Project::insert_resource(
            project_id,
            diesel::insert_into(dsl::snapshot)
                .values(snapshot)
                .on_conflict((dsl::project_id, dsl::name))
                .filter_target(dsl::time_deleted.is_null())
                .do_nothing(),
        )
        .insert_and_get_results_async(self.pool_authorized(opctx).await?)
        .await
        .map_err(|e| match e {
            AsyncInsertError::CollectionNotFound => Error::ObjectNotFound {
                type_name: ResourceType::Project,
                lookup_type: LookupType::ById(project_id),
            },
            AsyncInsertError::DatabaseError(e) => {
                public_error_from_diesel_pool(e, ErrorHandler::Server)
            }
        })?;
    let Some(snapshot) = snapshots.get(0) else {
         // Return an error -- we failed to insert the snapshot.
    };
```

Reply:
I originally tried `insert_and_get_result_async` with the additional `filter` (once I got it working), but I saw that it returned `CollectionNotFound` instead of `DatabaseError`. I'm trying the second suggestion now.